File: pkg/scene/cluster.go
package scene

type Cluster struct {
	id       ClusterID
	name     string
	property PropertyID
}

func NewCluster(cid ClusterID, name string, pid PropertyID) (*Cluster, error) {
	if cid.IsNil() {
		return nil, ErrInvalidID
	}
	return &Cluster{
		id:       cid,
		name:     name,
		property: pid,
	}, nil
}

func (c *Cluster) ID() ClusterID {
	if c == nil {
		return ClusterID{}
	}
	return c.id
}

func (c *Cluster) Name() string {
	if c == nil {
		return ""
	}
	return c.name
}

func (c *Cluster) Property() PropertyID {
	if c == nil {
		return PropertyID{}
	}
	return c.property
}

func (c *Cluster) Rename(name string) {
	if c == nil {
		return
	}
	c.name = name
}

func (c *Cluster) UpdateProperty(pid PropertyID) {
	if c == nil {
		return
	}
	c.property = pid
}

File: web/src/services/gql/queries/cluster.ts
import { gql } from "@apollo/client";

import { propertyFragment } from "@reearth/gql/fragments";

export const GET_CLUSTERS = gql`
  query GetClusters($sceneId: ID!, $lang: Lang) {
    node(id: $sceneId, type: SCENE) {
      id
      ... on Scene {
        clusters {
          id
          name
          propertyId
          property {
            id
            ...PropertyFragment
          }
        }
      }
    }
  }
`;

export const ADD_CLUSTER = gql`
  mutation AddCluster($sceneId: ID!, $name: String!, $lang: Lang) {
    addCluster(input: { sceneId: $sceneId, name: $name }) {
      cluster {
        id
        name
        propertyId
        property {
          id
          ...PropertyFragment
        }
      }
    }
  }

  ${propertyFragment}
`;

export const REMOVE_CLUSTER = gql`
  mutation RemoveCluster($sceneId: ID!, $clusterId: ID!) {
    removeCluster(input: { sceneId: $sceneId, clusterId: $clusterId }) {
      scene {
        id
        clusters {
          id
          name
          propertyId
        }
      }
    }
  }
`;

export const UPDATE_CLUSTER = gql`
  mutation UpdateCluster($sceneId: ID!, $clusterId: ID!, $name: String!) {
    updateCluster(input: { sceneId: $sceneId, clusterId: $clusterId, name: $name }) {
      cluster {
        id
        name
        propertyId
      }
    }
  }
`;
